python/tskit/metadata.py

```diff
@@ -109,13 +110,33 @@ def register_metadata_codec(
 
 class JSONCodec(AbstractMetadataCodec):
     def __init__(self, schema: Mapping[str, Any]) -> None:
-        pass
+        # Recurse through the schema finding default values to fill in on decode
+        self.defaults = {}
+
+        def record_defaults(path, sub_schema):
+            if "object" in sub_schema.get("type", "object"):
+                for key, prop in sub_schema.get("properties", {}).items():
+                    record_defaults(path + (key,), prop)
+            if "default" in sub_schema:
+                self.defaults[path] = sub_schema["default"]
+
+        record_defaults((), schema)
 
     def encode(self, obj: Any) -> bytes:
         return tskit.canonical_json(obj).encode()
 
     def decode(self, encoded: bytes) -> Any:
-        return json.loads(encoded.decode())
+        result = json.loads(encoded.decode())
+        # Assign default values
```

Comment:
I think it's logically simpler to start with the ``defaults`` dictionary, and write in anything that's in the decoded version, so
```python
result = copy.deepcopy(self.defaults)
recursive_update(result, decoded)
```

I agree this is made substantially trickier by the nestedness requirements.

I'm tempted to just implement this for top-level values initially (throwing an error if someone tries to set a default for a nested type), and see how far it gets us. It'll be a lot more work up front if we decide to support all possible combinations of nested defaults.